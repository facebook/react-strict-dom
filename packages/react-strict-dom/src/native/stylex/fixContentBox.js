/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow strict
 */

import type { ReactNativeStyle } from '../../types/renderer.native';

import { warnMsg } from '../../shared/logUtils';

type Direction = 0 | 1 | 2 | 3;

const TOP: Direction = 0;
const RIGHT: Direction = 1;
const BOTTOM: Direction = 2;
const LEFT: Direction = 3;

// TODO: check for RTL and flip START and END
const START: Direction = LEFT;
const END: Direction = RIGHT;

const borderMapping: [string, Direction[]][] = [
  ['borderWidth', [TOP, RIGHT, BOTTOM, LEFT]],
  ['borderTopWidth', [TOP]],
  ['borderRightWidth', [RIGHT]],
  ['borderBottomWidth', [BOTTOM]],
  ['borderLeftWidth', [LEFT]],
  ['borderStartWidth', [START]],
  ['borderEndWidth', [END]]
];

const paddingMapping: [string, Direction[]][] = [
  ['padding', [TOP, RIGHT, BOTTOM, LEFT]],
  ['paddingVertical', [TOP, BOTTOM]],
  ['paddingHorizontal', [LEFT, RIGHT]],
  ['paddingTop', [TOP]],
  ['paddingRight', [RIGHT]],
  ['paddingBottom', [BOTTOM]],
  ['paddingLeft', [LEFT]],
  ['paddingStart', [START]],
  ['paddingEnd', [END]]
];

export function fixContentBox(flatStyle: ReactNativeStyle): ReactNativeStyle {
  const border: [number, number, number, number] = [0, 0, 0, 0];
  const padding: [number, number, number, number] = [0, 0, 0, 0];

  for (const [styleProp, directions] of borderMapping) {
    if (typeof flatStyle[styleProp] === 'number') {
      for (const direction of directions) {
        border[direction] = flatStyle[styleProp];
      }
    }
  }

  for (const [styleProp, directions] of paddingMapping) {
    if (typeof flatStyle[styleProp] === 'number') {
      for (const direction of directions) {
        padding[direction] = flatStyle[styleProp];
      }
    }
  }

  const correctionVertical =
    border[TOP] + padding[TOP] + padding[BOTTOM] + border[BOTTOM];
  const correctionHorizontal =
    border[LEFT] + padding[LEFT] + padding[RIGHT] + border[RIGHT];

  const correctionMapping = new Map<string, number>([
    ['width', correctionHorizontal],
    ['minWidth', correctionHorizontal],
    ['maxWidth', correctionHorizontal],
    ['height', correctionVertical],
    ['minHeight', correctionVertical],
    ['maxHeight', correctionVertical]
  ]);

  const nextStyle: ReactNativeStyle = {};
  for (const styleProp of Object.keys(flatStyle)) {
    const correction = correctionMapping.get(styleProp);
    const styleValue = flatStyle[styleProp];
    if (correction != null) {
      if (styleValue == null) {
        nextStyle[styleProp] = null;
        continue;
      }
      if (styleValue === 'auto') {
        nextStyle[styleProp] = styleValue;
        continue;
      }
      if (typeof styleValue !== 'number') {
        if (__DEV__) {
          warnMsg(
            `unsupported style value in "${styleProp}:${String(
              styleValue
            )}" when used with "boxSizing:'content-box'". Expected a value that resolves to a number. Percentage values can only be used with "boxSizing:'border-box'".`
          );
        }
        return flatStyle;
      }
      nextStyle[styleProp] = styleValue + correction;
    } else {
      if (styleProp !== 'boxSizing') {
        nextStyle[styleProp] = styleValue;
      }
    }
  }

  return nextStyle;
}
