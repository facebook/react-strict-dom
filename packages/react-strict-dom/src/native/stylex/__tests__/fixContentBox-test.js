/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import { fixContentBox } from '../fixContentBox';

describe('fixContentBox', () => {
  beforeEach(() => {
    jest.spyOn(console, 'warn');
    console.warn.mockImplementation(() => {});
  });

  afterEach(() => {
    console.warn.mockRestore();
  });

  test('boxSizing: content-box', () => {
    const styles = {
      width: {
        boxSizing: 'content-box',
        borderWidth: 2,
        padding: 10,
        width: 100,
        // test that unrelated properties are unmodified
        overflow: 'hidden'
      },
      height: {
        boxSizing: 'content-box',
        borderWidth: 2,
        padding: 10,
        height: 50
      },
      maxWidth: {
        boxSizing: 'content-box',
        borderWidth: 2,
        padding: 10,
        maxWidth: 100
      },
      minWidth: {
        boxSizing: 'content-box',
        borderWidth: 2,
        padding: 10,
        minWidth: 100
      },
      maxHeight: {
        boxSizing: 'content-box',
        borderWidth: 2,
        padding: 10,
        maxHeight: 50
      },
      minHeight: {
        boxSizing: 'content-box',
        borderWidth: 2,
        padding: 10,
        minHeight: 50
      },
      allDifferent: {
        boxSizing: 'content-box',
        borderTopWidth: 1,
        borderRightWidth: 2,
        borderBottomWidth: 3,
        borderLeftWidth: 4,
        paddingTop: 10,
        paddingRight: 20,
        paddingBottom: 30,
        paddingLeft: 40,
        width: 100,
        height: 100
      },
      auto: {
        boxSizing: 'content-box',
        borderWidth: 2,
        padding: 10,
        height: 50,
        width: 'auto'
      },
      null: {
        boxSizing: 'content-box',
        borderWidth: 2,
        padding: 10,
        height: 50,
        width: null
      },
      string: {
        boxSizing: 'content-box',
        borderWidth: 2,
        padding: 10,
        height: 50,
        width: '50%'
      }
    };

    expect(fixContentBox(styles.width)).toMatchSnapshot('width');
    expect(fixContentBox(styles.height)).toMatchSnapshot('height');
    expect(fixContentBox(styles.maxWidth)).toMatchSnapshot('maxWidth');
    expect(fixContentBox(styles.maxHeight)).toMatchSnapshot('maxHeight');
    expect(fixContentBox(styles.minWidth)).toMatchSnapshot('minWidth');
    expect(fixContentBox(styles.minHeight)).toMatchSnapshot('minHeight');
    expect(fixContentBox(styles.allDifferent)).toMatchSnapshot('allDifferent');
    expect(fixContentBox(styles.auto)).toMatchSnapshot('auto');
    expect(fixContentBox(styles.null)).toMatchSnapshot('null');
    expect(fixContentBox(styles.string)).toMatchSnapshot('string');

    expect(console.warn).toHaveBeenCalledTimes(1);
  });
});
