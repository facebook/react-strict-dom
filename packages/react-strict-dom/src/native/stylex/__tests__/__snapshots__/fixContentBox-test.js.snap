// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fixContentBox boxSizing: content-box: allDifferent 1`] = `
{
  "borderBottomWidth": 3,
  "borderLeftWidth": 4,
  "borderRightWidth": 2,
  "borderTopWidth": 1,
  "height": 144,
  "paddingBottom": 30,
  "paddingLeft": 40,
  "paddingRight": 20,
  "paddingTop": 10,
  "width": 166,
}
`;

exports[`fixContentBox boxSizing: content-box: auto 1`] = `
{
  "borderWidth": 2,
  "height": 74,
  "padding": 10,
  "width": "auto",
}
`;

exports[`fixContentBox boxSizing: content-box: height 1`] = `
{
  "borderWidth": 2,
  "height": 74,
  "padding": 10,
}
`;

exports[`fixContentBox boxSizing: content-box: maxHeight 1`] = `
{
  "borderWidth": 2,
  "maxHeight": 74,
  "padding": 10,
}
`;

exports[`fixContentBox boxSizing: content-box: maxWidth 1`] = `
{
  "borderWidth": 2,
  "maxWidth": 124,
  "padding": 10,
}
`;

exports[`fixContentBox boxSizing: content-box: minHeight 1`] = `
{
  "borderWidth": 2,
  "minHeight": 74,
  "padding": 10,
}
`;

exports[`fixContentBox boxSizing: content-box: minWidth 1`] = `
{
  "borderWidth": 2,
  "minWidth": 124,
  "padding": 10,
}
`;

exports[`fixContentBox boxSizing: content-box: null 1`] = `
{
  "borderWidth": 2,
  "height": 74,
  "padding": 10,
  "width": null,
}
`;

exports[`fixContentBox boxSizing: content-box: string 1`] = `
{
  "borderWidth": 2,
  "boxSizing": "content-box",
  "height": 50,
  "padding": 10,
  "width": "50%",
}
`;

exports[`fixContentBox boxSizing: content-box: width 1`] = `
{
  "borderWidth": 2,
  "overflow": "hidden",
  "padding": 10,
  "width": 124,
}
`;
